spec/app/ics-031-crosschain-queries/README.md

```diff
@@ -0,0 +1,388 @@
+
+```typescript
+enum QueryResult {
+  SUCCESS,
+  FAILURE,
+  TIMEOUT,
+}
+```
+- A query that returns a value is marked as `SUCCESS`. This means that the query has been executed at the Queried Chain and there was a value associated to the queried path at the requested height.
+- A query that is executed but does not return a value is marked as `FAILURE`. This means that the query has been executed at the Queried Chain, but there was no value associated to the queried path at the requested height.
+- A query that timed out before a result is committed at the Querying Chain is marked as `TIMEOUT`.
+
+A CrossChainQueryResult is a particular interface used to represent query results.
+
+```typescript
+interface CrossChainQueryResult struct {
+    id: Identifier
+    result: QueryResult
+    data: []byte
+}
+```
+
+- The `id` field uniquely identifies the query at the Querying Chain.
+- The `result` field indicates whether the query was correctly executed at the Queried Chain and if the queried path exists.
+- The `data` field is an opaque bytestring that contains the value associated with the queried path in case `result = SUCCESS`.
+
+### Store paths
+
+#### Query path
+
+The query path is a private path that stores the state of ongoing cross-chain queries.
+
+```typescript
+function queryPath(id: Identifier): Path {
+    return "queries/{id}"
+}
+```
+#### Result query path
+
+The result query path is a public path that stores the result of completed queries.
+
+```typescript
+function resultQueryPath(id: Identifier): Path {
+    return "queriesresult/{id}"
+}
+```
+
+### Helper functions
+
+The Querying Chain MUST implement a function `generateQueryIdentifier`, which generates a unique query identifier:
+
+```typescript
+function generateQueryIdentifier = () -> Identifier
+```
+
+### Sub-protocols
+
+#### Query lifecycle
+
+1) When the Querying Chain receives a query request, it calls `CrossChainQueryRequest` of the Cross-chain Querying module. This function generates a unique identifier for the query, stores it in its `privateStore` and emits a `sendQuery` event. Query requests can be submitted by other IBC modules as transactions to the Querying Chain or simply executed as part of the `BeginBlock` and `EndBlock` logic.
+2) A correct relayer listening to `sendQuery` events from the Querying Chain will eventually pick the query request up and execute it at the Queried Chain. The result is then submitted (on-chain) to the Querying Chain.
+3) When the query result is committed at the Querying Chain, this calls the `CrossChainQueryResult` function of the Cross-chain Querying module.
+4) The `CrossChainQueryResult` first retrieves the query from the `privateStore` using the query's unique identifier. It then proceeds to verify the result using its local client. If it passes the verification, the function removes the query from the `privateStore` and stores the result in a public path.
```

Comment:
Why does the result need to be in a public path?

Reply:
No need. Fixed.